### PR TITLE
Feature/attribs on params

### PR DESCRIPTION
**Pull request by kevinwright**
Added test for annotation capture
Updated to lamdafied syntax for SAM type construction
Minor changes to permit compilation under JDK 9
Added Kevin Wright as a contributor

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -27,13 +27,21 @@ lazy val tests = project
   .settings(moduleName := "magnolia-tests")
   .settings(
     addCompilerPlugin("org.scalamacros" % "paradise" % "2.1.1" cross CrossVersion.full),
+    initialCommands in console := """import magnolia.tests._; import magnolia.examples._;""",
```
Comment (propensive) [inline build.sbt]: Nice. :)

File: build.sbt [hunk 2]
```diff
@@ -27,13 +27,21 @@ lazy val tests = project
...
     libraryDependencies ++= Seq(
       "com.fommil" %% "deriving-macro" % "0.9.0",
-      "com.fommil" %% "scalaz-deriving" % "0.9.0"
+      "com.fommil" %% "scalaz-deriving" % "0.9.0",
```
Comment (propensive) [inline build.sbt]: Looks like `scalaz-deriving` crept back in... can we drop it?

File: build.sbt [hunk 3]
```diff
@@ -84,12 +92,15 @@ lazy val publishSettings = Seq(
         <name>Jon Pretty</name>
         <url>https://github.com/propensive/magnolia/</url>
       </developer>
+      <developer>
```
Comment (propensive) [inline build.sbt]: I've been using the `CONTRIBUTORS` file to manage this, though I'm not sure what is convention, or best practice, except that using a combination of both almost certainly isn't. Any ideas?

File: core/shared/src/main/scala/interface.scala [hunk 4]
```diff
@@ -25,6 +25,8 @@ trait Subtype[Typeclass[_], Type] {
 
   /** partial function defined the subset of values of `Type` which have the type of this subtype */
   def cast: PartialFunction[Type, SType]
+
+  override def toString: String = s"Subtype(${typeName.full})"
```
Comment (propensive) [inline core/shared/src/main/scala/interface.scala]: Very welcome. :)

File: core/shared/src/main/scala/interface.scala [hunk 5]
```diff
@@ -128,7 +142,7 @@ abstract class CaseClass[Typeclass[_], Type] private[magnolia] (
 
   /** constructs a new instance of the case class type
     *
-    *  Like [[construct]] this method is implemented by Magnolia and let's you construct case class
+    *  Like [[construct]] this method is implemented by Magnolia and lets you construct case class
```
Comment (propensive) [inline core/shared/src/main/scala/interface.scala]: Hanging my head in shame.

File: core/shared/src/main/scala/interface.scala [hunk 6]
```diff
@@ -145,20 +159,30 @@ abstract class CaseClass[Typeclass[_], Type] private[magnolia] (
...
 }
 
 /** represents a sealed trait and the context required to construct a new typeclass instance
   *  corresponding to it
   *
   *  Instances of `SealedTrait` provide access to all of the component subtypes of the sealed trait
   *  which form a coproduct, and to the fully-qualified name of the sealed trait.
-  *
   *  @param typeName       the name of the sealed trait
   *  @param subtypesArray  an array of [[Subtype]] instances for each subtype in the sealed trait
```
Comment (propensive) [inline core/shared/src/main/scala/interface.scala]: Add `@param annotationsArray ...` in here.

File: core/shared/src/main/scala/magnolia.scala [hunk 7]
```diff
@@ -238,34 +255,39 @@ object Magnolia {
...
-          // If a companion object is defined with alternative apply methods
-          // it is needed get all the alternatives
-          val constructorMethods =
-            companionSym.decl(TermName("apply")).alternatives.map(_.asMethod)
 
-          // The last apply method in the alternatives is the one that belongs
-          // to the case class, not the user defined companion object
-          val indexedConstructorParams =
-            constructorMethods.last.paramLists.head.map(_.asTerm).zipWithIndex
+        val defaults = headParamList.filterNot(_ => isValueClass) map { plist =>
```
Comment (propensive) [inline core/shared/src/main/scala/magnolia.scala]: The `filterNot` on an external value looks a bit unconventional... Not hugely fussed about it, but how about just `if(isValueClass) Nil else ...`?

File: examples/shared/src/main/scala/decode.scala [hunk 8]
```diff
@@ -10,12 +10,10 @@ trait Decoder[T] { def decode(str: String): T }
...
 
   /** decodes strings */
-  implicit val string: Decoder[String] = new Decoder[String] {
-    def decode(str: String): String = str
-  }
+  implicit val string: Decoder[String] = (str: String) => str
 
   /** decodes ints */
-  implicit val int: Decoder[Int] = new Decoder[Int] { def decode(str: String): Int = str.toInt }
+  implicit val int: Decoder[Int] = (str: String) => str.toInt
```
Comment (propensive) [inline examples/shared/src/main/scala/decode.scala]: Do you actually need the `: String` type annotation? I think `_.toInt` should just work.

File: examples/shared/src/main/scala/decode.scala [hunk 9]
```diff
@@ -24,22 +22,18 @@ object Decoder {
   type Typeclass[T] = Decoder[T]
 
   /** defines how new [[Decoder]]s for case classes should be constructed */
-  def combine[T](ctx: CaseClass[Decoder, T]): Decoder[T] = new Decoder[T] {
-    def decode(value: String) = {
-      val (_, values) = parse(value)
-      ctx.construct { param =>
-        param.typeclass.decode(values(param.label))
-      }
+  def combine[T](ctx: CaseClass[Decoder, T]): Decoder[T] = (value: String) => {
```
Comment (propensive) [inline examples/shared/src/main/scala/decode.scala]: ...and here.

File: examples/shared/src/main/scala/show.scala [hunk 10]
```diff
@@ -20,26 +20,30 @@ trait GenericShow[Out] {
...
     *  showing each parameter, and prefixing it with the class name */
-  def combine[T](ctx: CaseClass[Typeclass, T]): Show[Out, T] = new Show[Out, T] {
-    def show(value: T) =
-      if (ctx.isValueClass) {
-        val param = ctx.parameters.head
-        param.typeclass.show(param.dereference(value))
-      } else {
-        val paramStrings = ctx.parameters.map { param =>
-          s"${param.label}=${param.typeclass.show(param.dereference(value))}"
+  def combine[T](ctx: CaseClass[Typeclass, T]): Show[Out, T] = (value: T) => {
```
Comment (propensive) [inline examples/shared/src/main/scala/show.scala]: My personal preference would be to write this as literally, `= { value =>`, not `= value => {`. (Removing the type ascription, too.)

File: examples/shared/src/main/scala/show.scala [hunk 11]
```diff
@@ -20,26 +20,30 @@ trait GenericShow[Out] {
...
+    if (ctx.isValueClass) {
+      val param = ctx.parameters.head
+      param.typeclass.show(param.dereference(value))
+    } else {
+      val paramStrings = ctx.parameters.map { param =>
+        val attribStr = if(param.annotations.isEmpty) "" else {
```
Comment (propensive) [inline examples/shared/src/main/scala/show.scala]: Nice addition!

File: examples/shared/src/main/scala/show.scala [hunk 12]
```diff
@@ -49,17 +53,13 @@ trait GenericShow[Out] {
 object Show extends GenericShow[String] {
 
   /** show typeclass for strings */
-  implicit val string: Show[String, String] = new Show[String, String] {
-    def show(s: String): String = s
-  }
+  implicit val string: Show[String, String] = (s: String) => s
```
Comment (propensive) [inline examples/shared/src/main/scala/show.scala]: `identity`? ;)

File: project/plugins.sbt [hunk 13]
```diff
@@ -1,5 +1,5 @@
-addSbtPlugin("com.jsuereth" % "sbt-pgp" % "1.1.0-M1")
-addSbtPlugin("org.xerial.sbt" % "sbt-sonatype" % "2.0")
+addSbtPlugin("com.jsuereth" % "sbt-pgp" % "1.1.0")
```
Comment (propensive) [inline project/plugins.sbt]: Build maintenance: much appreciated!

File: tests/src/main/scala/tests.scala [hunk 14]
```diff
@@ -1,12 +1,13 @@
 package magnolia.tests
 
 import language.experimental.macros
-import estrapade.{test, TestApp}
+import estrapade.{TestApp, test}
```
Comment (propensive) [inline tests/src/main/scala/tests.scala]: Sure, why not!

File: tests/src/main/scala/tests.scala [hunk 15]
```diff
@@ -253,10 +262,14 @@ object Tests extends TestApp {
...
         val person = Person("Bob", 42)
         implicitly[Patcher[Entity]].patch(person, Seq(null, 'killer))
+        "it worked"
       } catch {
         case NonFatal(e) => e.getMessage
       }
-    }.assert(_ == "scala.Symbol cannot be cast to java.lang.Integer")
+    }.assert{x =>
+      //tiny hack because Java 9 inserts the "java.base/" module name in the error message
+      x.startsWith("scala.Symbol cannot be cast to") && x.endsWith("java.lang.Integer")
```
Comment (propensive) [inline tests/src/main/scala/tests.scala]: That's horrible. But we're in test code now, so pragmatism reigns!

File: core/shared/src/main/scala/magnolia.scala (hunk 7 above)
Comment (kevinwright) [reply core/shared/src/main/scala/magnolia.scala]: Turns out that I can just lose the filter entirely, and all the tests pass just fine

File: examples/shared/src/main/scala/show.scala (hunk 12 above)
Comment (kevinwright) [reply examples/shared/src/main/scala/show.scala]: Nope, tried that...

File: tests/src/main/scala/tests.scala (hunk 14 above)
Comment (kevinwright) [reply tests/src/main/scala/tests.scala]: Blame intellij :)

File: examples/shared/src/main/scala/decode.scala (hunk 8 above)
Comment (kevinwright) [reply examples/shared/src/main/scala/decode.scala]: Good catch, I was being cautious after `identity` failed previously

File: build.sbt (hunk 3 above)
Comment (kevinwright) [reply build.sbt]: done

File: build.sbt (hunk 2 above)
Comment (kevinwright) [reply build.sbt]: done

File: examples/shared/src/main/scala/show.scala (hunk 11 above)
Comment (kevinwright) [reply examples/shared/src/main/scala/show.scala]: Had to be able to test it somehow :)

File: examples/shared/src/main/scala/show.scala (hunk 12 above)
Comment (propensive) [reply examples/shared/src/main/scala/show.scala]: 😢

**Pull request merged**